MLflow autologging issue # 1618

dennisbader opened the pull request:
Addresses #1618.

## Summary

Draft PR for  MLflow integration example making pytorch calls and activating tensorboard collection on Darts' s pytortch models to do autologging with MLflow and activate MLflow UI.

cargecla1 committed:
MLflow autologging issue # 1618

MLflow integration example making pytorch calls and activating tensorboard collection on Darts' s pytortch models to do autologging with MLflow and activate MLflow UI.

cargecla1 commented:
Hello @dennisbader,

As a follow up, is there anything else that I need to do? 
What is/are the next step/steps?

Cheers!

dennisbader commented:
Hi @cargecla1. No, all good from your side 🚀 
Right now our capacity is a bit limited before the holidays, so we'll most likely have time to review this early next year.
Thanks again for the contribution 💯

cargecla1 commented:
> Hi @cargecla1. No, all good from your side 🚀 Right now our capacity is a bit limited before the holidays, so we'll most likely have time to review this early next year. Thanks again for the contribution 💯

Hello @dennisbader,

Thanks for letting me know.
And thank you for supporting this review.

Merry Xmas!

cargecla1 reviewed (approved):
I agree with this changes, there are as passed on.

turbotimon commented on docs/userguide/torch_forecasting_models.md:
```diff
@@ -461,6 +461,109 @@ model.fit(...)
+
+with mlflow.start_run(nested=True) as run:
+
+    dataset: PandasDataset = mlflow.data.from_pandas(series, source="AirPassengersDataset")
```
IMHO should call `series.pd_dataframe()` because in my tests it worked only on DataFrames not Series:
`dataset: PandasDataset = mlflow.data.from_pandas(series.pd_dataframe(), source="AirPassengersDataset")`

cargecla1 replied:
Good suggestion

cargecla1 replied:
Added via commit c15eb1e9244d29cca103d81700f2d9ecf6c5ecf4
Thank you!

turbotimon commented on docs/userguide/torch_forecasting_models.md:
````diff
@@ -461,6 +461,109 @@ model.fit(...)
+MLflow using interface (UI) and autologging to track Dart's pytorch models. 
+```python
+import pandas as pd
+from torchmetrics import MeanAbsolutePercentageError
````
`import torchmetrics` as later torchmetrics is used directly (L486)

cargecla1 replied:
Good idea.

cargecla1 replied:
Added via commit c15eb1e9244d29cca103d81700f2d9ecf6c5ecf4
Thanks!

turbotimon commented on docs/userguide/torch_forecasting_models.md:
```diff
@@ -461,6 +461,109 @@ model.fit(...)
+
+# Registering model
+model_uri = f"runs:/{run.info.run_id}/darts-NBEATS"
+mlflow.register_model(model_uri=model_uri, name=model_name)
```
IMHO runs currently do not save models as artifacts becaus there is no call to `log_model()`. Therefore `model_uri` is not pointing to a valid model. Please correct me if I'm wrong, however, it did not work in my tests. If it works, can you pleas add an example how to load with ` loaded_model = mlflow.<flavor>.load_model(model_uri=model_uri)` ?

turbotimon replied:
As i understand it you can only (kind of) promote a already saved model to a registered model..

> Model
> An MLflow Model is created from an experiment or run that is logged with one of the model flavor’s mlflow.<model_flavor>.log_model() methods. **Once logged, this model can then be registered with the Model Registry.**
> 
> Registered Model
> An MLflow Model can be registered with the Model Registry. A registered model has a unique name, contains versions, aliases, tags, and other metadata.
https://mlflow.org/docs/latest/model-registry.html#concepts

cargecla1 replied:
> As i understand it you can only (kind of) promote a already saved model to a registered model..
> 
> > Model
> > An MLflow Model is created from an experiment or run that is logged with one of the model flavor’s mlflow.<model_flavor>.log_model() methods. **Once logged, this model can then be registered with the Model Registry.**
> > Registered Model
> > An MLflow Model can be registered with the Model Registry. A registered model has a unique name, contains versions, aliases, tags, and other metadata.
> > https://mlflow.org/docs/latest/model-registry.html#concepts

Hello @turbotimon ,
My plans was to share how to do the first part of issue # 1618. As it allows you to track the experiments (so one don't get lost if you are doing more of a few iterations), select the best one, see metrics progress, etc.
I will have a look at what you are suggesting. However, the normal mlflow.python.loadmodel or mlflow.pytorch.loadmodel doesn't seems to work because the way Darts is wrapping torch.nn modules. All test I have done so far didn't work to save and load  the model. Only to record hyperparameters, other torchmetrics related information, dataset hash, etc. Thus, I propose to share this, and later we can add saving and loading models (but my last weekend research it is pointing out that the model.py files and common models may need to be rewritten to be compatible with MLFlow (I might be wrong).

Cheers!

turbotimon replied:
@cargecla1 Yes, that's a good idea to split these two things

If using the mlflow model registry for darts completely fails, you could also mentioning the workaround I proposed in the [issue](https://github.com/unit8co/darts/issues/1618#issuecomment-1753125634). Which was manually saving/loading the model as an artifact. Something like:
```python
dartsmodel.save("mymodel.pickle")
mlflow.log_artifact("mymodel.pickle")
# later, load/download artifact from mlflow and do e.g. a RNNModel.load("mymodel.pickle")
```

Let me know if i can help anything!

cargecla1 replied:
> @cargecla1 Yes, that's a good idea to split these two things
> 
> If using the mlflow model registry for darts completely fails, you could also mentioning the workaround I proposed in the [issue](https://github.com/unit8co/darts/issues/1618#issuecomment-1753125634). Which was manually saving/loading the model as an artifact. Something like:
> 
> ```python
> dartsmodel.save("mymodel.pickle")
> mlflow.log_artifact("mymodel.pickle")
> # later, load artifact from mlflow and do e.g. a dartsmodel=RNNModel.load("mymodel.pickle")
> ```
> 
> Let me know if i can help anything!

Hello @turbotimon ,

Copy that, I forgot this comments from the issue you linked on your message.
This could work because you can save a darts model using pickle (and mlflow) and you can load models using pickle API.
Additional references to try:
https://mlflow.org/docs/latest/model-registry.html?highlight=pickle#id18
https://mlflow.org/docs/latest/model-registry.html?highlight=pickle#id19

Thank you!

cla-ra3426 replied:
Hello @turbotimon 

After trying to save the Darts model using dartsmodel.save("mymodel.pickle") as suggested, I am getting the following error:

File ~\.conda\envs\venv_darts\lib\site-packages\torch\serialization.py:653, in _save(obj, zip_file, pickle_module, pickle_protocol)
    651 pickler = pickle_module.Pickler(data_buf, protocol=pickle_protocol)
    652 pickler.persistent_id = persistent_id
--> 653 pickler.dump(obj)
    654 data_value = data_buf.getvalue()
    655 zip_file.write_record('data.pkl', data_value, len(data_value))

AttributeError: Can't pickle local object 'LayerSummary._register_hook.<locals>.hook'

Would you know why this is happening?

Cheers

cla-ra3426 replied:
> @cargecla1 Yes, that's a good idea to split these two things
> 
> If using the mlflow model registry for darts completely fails, you could also mentioning the workaround I proposed in the [issue](https://github.com/unit8co/darts/issues/1618#issuecomment-1753125634). Which was manually saving/loading the model as an artifact. Something like:
> 
> ```python
> dartsmodel.save("mymodel.pickle")
> mlflow.log_artifact("mymodel.pickle")
> # later, load artifact from mlflow and do e.g. a dartsmodel=RNNModel.load("mymodel.pickle")
> ```
> 
> Let me know if i can help anything!

Hello @turbotimon , @madtoinou , @dennisbader,

Is it possible to split this issue as discussed above? so we can share how to use, train, track, monitor and save the models using MLFlow? Just leaving loading the model for a later release?

Cheers

turbotimon replied:
> Hello @turbotimon
> 
> After trying to save the Darts model using dartsmodel.save("mymodel.pickle") as suggested, I am getting the following error:
> 
> File ~.conda\envs\venv_darts\lib\site-packages\torch\serialization.py:653, in _save(obj, zip_file, pickle_module, pickle_protocol) 651 pickler = pickle_module.Pickler(data_buf, protocol=pickle_protocol) 652 pickler.persistent_id = persistent_id --> 653 pickler.dump(obj) 654 data_value = data_buf.getvalue() 655 zip_file.write_record('data.pkl', data_value, len(data_value))
> 
> AttributeError: Can't pickle local object 'LayerSummary._register_hook..hook'
> 
> Would you know why this is happening?
> 
> Cheers

Hi @cla-ra3426, sorry i missed your question. No idea.. but must have to do somehting with darts itself and not mlflow i suppose

cargecla1 replied:
> > @cargecla1 Yes, that's a good idea to split these two things
> > If using the mlflow model registry for darts completely fails, you could also mentioning the workaround I proposed in the [issue](https://github.com/unit8co/darts/issues/1618#issuecomment-1753125634). Which was manually saving/loading the model as an artifact. Something like:
> > ```python
> > dartsmodel.save("mymodel.pickle")
> > mlflow.log_artifact("mymodel.pickle")
> > # later, load artifact from mlflow and do e.g. a dartsmodel=RNNModel.load("mymodel.pickle")
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Let me know if i can help anything!
> 
> Hello @turbotimon , @madtoinou , @dennisbader,
> 
> Is it possible to split this issue as discussed above? so we can share how to use, train, track, monitor and save the models using MLFlow? Just leaving loading the model for a later release?
> 
> Cheers

Hello gents (@turbotimon , @madtoinou , @dennisbader,)

Have you had time to consider my proposal above?  _"Is it possible to split this issue as discussed above? so we can share how to use, train, track, monitor and save the models using MLFlow? Just leaving loading the model for a later release?"_

The only solution I have right now it is to train the model with MLFlow to track and monitor the model and retrain it with pure Darts to be able to save it and load to predict, Darts saving method "fails" when run inside a MLFlow run, and MLFlow log and save methods don't want to work with Darts either.

Thank you in advice for your consideration!

Cheers!

madtoinou replied:
Hi @cargecla1,

Sorry for the delay, loading model is kind-of part of linked issue so it would probably be better to also include it in this PR. But if it too troublesome, we can treat it separately. 

@cla-ra3426,

The error you're getting when trying to pickle the model is probably due to the (pytorch-lightning) callbacks. Can you try removing them prior to exporting the model?

I will test this example more thoroughly when I have more time, try to see if I can come up with a solution for the loading.

cargecla1 replied:
Hello @madtoinou 

I will try this and see how it goes.

Thanks for this suggestion and for considering splitting the problem into two if above doesn't work.

cargecla1 replied:
Hello @madtoinou ,

I took your suggestion on board as per above, but that didn't fixed the issue with loading and predicting steps.

Refer to commit https://github.com/unit8co/darts/pull/2092/commits/f7982440b876ff6d07761a2ed6ac8717fe93a1c0

cargecla1 committed:
Update torch_forecasting_models.md

Adding first two suggestions raised in MLflow autologging issue # 1618 #2092 via comments.

cargecla1 committed:
Merge branch 'master' into feature/MLflow_1618

cargecla1 committed:
Saving pip and conda env

Adding code to save pip requirements and conda environment required to run the model via MLFlow.

cargecla1 committed:
Track and save model

Adding a way to track and save model before registering it using MLFlow.

cargecla1 committed:
Expanding comment

Expanding comment to create a new cell to log the artifact and register the model after a model run id has been created.

madtoinou commented on docs/userguide/torch_forecasting_models.md:
```diff
@@ -461,6 +461,122 @@ model.fit(...)
 
 *Note* : The callback will give one more element in the `loss_logger.val_loss` as the model trainer performs a validation sanity check before the training begins.
 
+#### Example with MLflow Autologging
```
Can you please also add an entry in the table of content at the top of the file/page?

cargecla1 replied:
Done via commit https://github.com/unit8co/darts/pull/2092/commits/ce635bc838f2bc187a9b1cff18f029df45bd7265

madtoinou commented on docs/userguide/torch_forecasting_models.md:
```diff
@@ -461,6 +461,122 @@ model.fit(...)
+    # dataset is used for model training
+    mlflow.log_input(dataset, context="training")
+   
+    mlflow.log_param("model_type", "Darts_Pytorch_model")
+    mlflow.log_param("input_chunk_length", 24)
+    mlflow.log_param("output_chunk_length", 12)                   
+    mlflow.log_param("n_epochs", 500)
+    mlflow.log_param("model_name", 'NBEATS_MLflow')
+    mlflow.log_param("log_tensorboard", True)
+    mlflow.log_param("torch_metrics", "torchmetrics.regression.MeanAbsolutePercentageError()")
+    mlflow.log_param("nr_epochs_val_period", 1)
+    mlflow.log_param("pl_trainer_kwargs", "{callbacks: [loss_logger]}")    
```
Let's make this a bit more concise, with a list of parameters to log and access them in the `model.model_params`, `model.pl_module_params` or `model.trainer_params` attributes (potentially after fusing these dictionaries into one).

cargecla1 replied:
Hello @madtoinou 
I made the code more concise as suggested. Refer to commit https://github.com/unit8co/darts/pull/2092/commits/f7982440b876ff6d07761a2ed6ac8717fe93a1c0

cargecla1 replied:
Hello @madtoinou , @dennisbader ,

I guess you may have been busy with the latest Darts release. Did you have time to look at my latest changes? Does this work to merge the first part of the MLFlow integration as previously discussed?

Cheers!

By the way, nicely done! I need to test the latest release speeds.  :)

cargecla1 committed:
Updating table of content

Updating table of content at the top of the file/page.

cargecla1 committed:
Making log more concise

Making the log parameters method more concise as suggested.

madtoinou commented on docs/userguide/torch_forecasting_models.md:
```diff
@@ -25,6 +25,7 @@ We assume that you already know about covariates in Darts. If you're new to the
     - [Callbacks](#callbacks)
       - [Early Stopping](#example-with-early-stopping)
       - [Custom Callback](#example-of-custom-callback-to-store-losses)
+      - [MLFlow: train, track and monitor](#example-with-mlflow-autologging)
```
I think that the indentation is off, can you shift it to the left once so that it's at the same level as "Callbacks"?

cargecla1 replied:
Hello @madtoinou ,

I fixed this via commit https://github.com/unit8co/darts/pull/2092/commits/b839c67ce2f970b23d3d9b41e9b4a0ab4ae35315

madtoinou commented on docs/userguide/torch_forecasting_models.md:
```diff
@@ -461,6 +462,109 @@ model.fit(...)
+    mlflow.log_params(params)    
+
+    # create the model
+    model = NBEATSModel(
+        input_chunk_length=24,
+        output_chunk_length=12,
+        n_epochs=500,
+        model_name='NBEATS_MLflow',
+        log_tensorboard=True,
+        torch_metrics=torch_metrics,
+        nr_epochs_val_period=1,
+        )
```
Instead of repeating the parameters, we could just unpack the `params` dictionary into the constructor (the type of `torch_metrics` might have to change?):
```suggestion
    model = NBEATSModel(
        **params,
        torch_metrics=torch_metrics,
        )
```

madtoinou reviewed (commented):
Thank you for making the correction, I still have some minor comments.

cargecla1 committed:
Indentation fix

Fixed indentation that was off.

cargecla1 committed:
Update docs/userguide/torch_forecasting_models.md

Adding suggestion, thanks!

Co-authored-by: madtoinou <32447896+madtoinou@users.noreply.github.com>

cargecla1 committed:
Merge branch 'master' into feature/MLflow_1618

cargecla1 committed:
Resolving torcmetrics

Resolving torchmetrics issue.

cargecla1 committed:
Update MLFlow saving method

Updating the MLFlow saving method including tracking uri.

cargecla1 commented:
> Thank you for making the correction, I still have some minor comments.

Hello @madtoinou ,

I just updated torchmetrics and saving model method to capture your latest suggestions with the **params. Hopefully this is work done now.   :)

Cheers!

cargecla1 commented:
> > Thank you for making the correction, I still have some minor comments.
> 
> Hello @madtoinou ,
> 
> I just updated torchmetrics and saving model method to capture your latest suggestions with the **params. Hopefully this is work done now. :)
> 
> Cheers!

Hello @madtoinou, @dennisbader ,

Any updates, comments from my latest updates/commits?

Cheers!

cargecla1 commented:
> > > Thank you for making the correction, I still have some minor comments.
> > 
> > 
> > Hello @madtoinou ,
> > I just updated torchmetrics and saving model method to capture your latest suggestions with the **params. Hopefully this is work done now. :)
> > Cheers!
> 
> Hello @madtoinou, @dennisbader ,
> 
> Any updates, comments from my latest updates/commits?
> 
> Cheers!

Hello @madtoinou , @dennisbader ,

Did you have chance to have a look a my latest changes to this PR?, any feedback will be appreciated.

Cheers!

cargecla1 committed:
Merge branch 'master' into feature/MLflow_1618

dennisbader commented:
Closing this in favor of #3022
